File: python/problems/leetcode/0239_sliding_window_maximum/solution.py
"""LeetCode 239: Sliding Window Maximum."""

from __future__ import annotations

from collections import deque


class Solution:
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        dq: deque[int] = deque()
        res: list[int] = []
        for i, x in enumerate(nums):
            while dq and nums[dq[-1]] <= x:
                dq.pop()
            dq.append(i)
            if dq[0] <= i - k:
                dq.popleft()
            if i >= k - 1:
                res.append(nums[dq[0]])
        return res


if __name__ == "__main__":
    assert Solution().maxSlidingWindow([1, 3, -1, -3, 5, 3, 6, 7], 3) == [
        3, 3, 5, 5, 6, 7,
    ]
    assert Solution().maxSlidingWindow([1], 1) == [1]
    print("0239 OK")

File: cpp/problems/leetcode/0239_sliding_window_maximum/solution.cpp
// LeetCode 239
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> dq;
        vector<int> res;
        for (int i = 0; i < (int)nums.size(); ++i) {
            while (!dq.empty() && nums[dq.back()] <= nums[i]) dq.pop_back();
            dq.push_back(i);
            if (dq.front() <= i - k) dq.pop_front();
            if (i >= k - 1) res.push_back(nums[dq.front()]);
        }
        return res;
    }
};

int main() {
    vector<int> a{1, 3, -1, -3, 5, 3, 6, 7};
    assert((Solution().maxSlidingWindow(a, 3) == vector<int>{3, 3, 5, 5, 6, 7}));
    vector<int> b{1};
    assert((Solution().maxSlidingWindow(b, 1) == vector<int>{1}));
    cout << "0239 OK" << endl;
    return 0;
}
